src/librustc_builtin_macros/asm.rs

```diff
@@ -485,34 +486,11 @@ fn expand_preparsed_asm(ecx: &mut ExtCtxt<'_>, sp: Span, args: AsmArgs) -> P<ast
         .into_iter()
         .enumerate()
         .filter(|&(_, used)| !used)
-        .map(|(idx, _)| {
-            if named_pos.contains(&idx) {
-                // named argument
-                (operands[idx].1, "named argument never used")
-            } else {
-                // positional argument
-                (operands[idx].1, "argument never used")
-            }
-        })
+        .map(|(idx, _)| operands[idx].1)
         .collect();
-    match unused_operands.len() {
-        0 => {}
-        1 => {
-            let (sp, msg) = unused_operands.into_iter().next().unwrap();
-            let mut err = ecx.struct_span_err(sp, msg);
-            err.span_label(sp, msg);
-            err.emit();
-        }
-        _ => {
-            let mut err = ecx.struct_span_err(
-                unused_operands.iter().map(|&(sp, _)| sp).collect::<Vec<Span>>(),
-                "multiple unused asm arguments",
-            );
-            for (sp, msg) in unused_operands {
-                err.span_label(sp, msg);
-            }
-            err.emit();
-        }
+    if !unused_operands.is_empty() {
+        let msg = format!("asm argument{} not used in template", pluralize!(unused_operands.len()));
+        ecx.parse_sess.buffer_lint(UNUSED_ASM_ARGUMENTS, unused_operands, ast::CRATE_NODE_ID, &msg);
```

Comment:
To suppress the warning `#![allow(unused_asm_arguments)]` will have to be crate-level due to `ast::CRATE_NODE_ID`, which is not good in this case.

We can do better, from the `asm` macro's expansion id `ecx.current_expansion.id` we can go to it's parent `LocalDefId` through `Definitions::invocation_parent`, and that `LocalDefId` can be converted to `NodeId`.
(`Definitions` should be available through `ecx.resolver`.)

It would be also good to have a test for an unused asm argument suppressed by a non crate-level `allow`.

Reply:
Unfortunately `Definitions` isn't available to librustc_builtin_macros since it is defined in librustc_hir. All I have is a `&mut dyn Resolver`.

One possible way around that would be to move the unused argument check to the AST lowering pass, thoughts?

Reply:
Data from `Definitions` is available through `dyn Resolver`, see how https://github.com/rust-lang/rust/pull/73178 does it.

These PRs probably don't need to block each other, but once the first one lands, the second one will need to be adjusted.